File: GLLara/GLLMeshDrawData.m
//
//  GLLMeshDrawData.m
//  GLLara
//
//  Created by Torsten Kammer on 01.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLMeshDrawData.h"

#import <OpenGL/gl3.h>

#import "NSArray+Map.h"
#import "GLLModelMesh.h"
#import "GLLModelProgram.h"
#import "GLLVertexFormat.h"
#import "GLLVertexArray.h"
#import "GLLUniformBlockBindings.h"
#import "GLLResourceManager.h"
#import "GLLTexture.h"

@interface GLLMeshDrawData ()
{
    GLLVertexArray *vertexArray;
}

@end

@implementation GLLMeshDrawData

@dynamic vertexArray;

- (id)initWithMesh:(GLLModelMesh *)mesh vertexArray:(GLLVertexArray *)array resourceManager:(GLLResourceManager *)resourceManager error:(NSError *__autoreleasing*)error;
{
    if (!(self = [super init])) return nil;
    
    _modelMesh = mesh;
    _elementsOrVerticesCount = (GLsizei) mesh.countOfElements;
    if (array.format.numElementBytes == 0) {
        // Rendering without elements
        _elementsOrVerticesCount = (GLsizei) mesh.countOfVertices;
    }
    vertexArray = array;
    _indicesStart = (GLsizeiptr) array.elementDataLength;
    _baseVertex = (GLint) array.countOfVertices;
    
    switch (array.format.numElementBytes) {
        case 4:
            _elementType = GL_UNSIGNED_INT;
            break;
        case 2:
            _elementType = GL_UNSIGNED_SHORT;
            break;
        case 1:
            _elementType = GL_UNSIGNED_BYTE;
            break;
        case 0:
            // Rendering without elements
            _elementType = 0;
            break;
        default:
            [NSException raise:NSInvalidArgumentException format:@"Can't deal with vertex format with %li bytes per index", array.format.numElementBytes];
    }
    
    [array addVertices:mesh.vertexDataAccessors count:mesh.countOfVertices elements:mesh.elementData elementsType:mesh.elementComponentType];
    
    return self;
}

- (void)unload
{
    vertexArray = nil;
    _elementsOrVerticesCount = 0;
}

- (void)dealloc
{
    NSAssert(vertexArray == 0 && _elementsOrVerticesCount == 0, @"Did not call unload before calling dealloc!");
}

- (GLuint)vertexArray
{
    return vertexArray.vertexArrayIndex;
}

- (NSComparisonResult)compareTo:(GLLMeshDrawData *)other;
{
    if (vertexArray != other->vertexArray) {
        return vertexArray < other->vertexArray ? NSOrderedAscending : NSOrderedDescending;
    }
    
    return NSOrderedSame;
}

@end

File: GLLara/GLLMeshDrawData.swift
//
//  GLLMeshDrawData.swift
//  GLLara
//
//  Created by Torsten Kammer on 10.04.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation
import Metal

class GLLMeshDrawData {
    
    let modelMesh: GLLModelMesh
    
    let elementType: MTLIndexType
    let baseVertex: Int
    let indicesStart: Int
    let elementsOrVerticesCount: Int
    let vertexArray: GLLVertexArray
    let boneDataArray: MTLBuffer?
    let boneIndexOffset: Int?
    let reservation: GLLVertexArray.Reservation
    
    init(mesh: GLLModelMesh, vertexArray array: GLLVertexArray, resourceManager: GLLResourceManager) {
        modelMesh = mesh
        self.vertexArray = array
        
        reservation = array.reserve(vertexCount: modelMesh.countOfVertices, elements: modelMesh.elementData, bytesPerElement: modelMesh.elementSize)
        
        indicesStart = reservation.elementBytesStart
        baseVertex = reservation.baseVertex
        elementType = vertexArray.format.indexType
        if vertexArray.format.hasIndices {
            elementsOrVerticesCount = mesh.countOfElements
        } else {
            elementsOrVerticesCount = mesh.countOfVertices
        }
        
        if let boneIndices = mesh.variableBoneIndices, let boneWeights = mesh.variableBoneWeights {
            let weightsSize = MemoryLayout<Float>.stride * boneWeights.count
            let indicesSize = MemoryLayout<UInt16>.stride * boneIndices.count
            
            boneDataArray = resourceManager.metalDevice.makeBuffer(length: weightsSize + indicesSize, options: .storageModeShared)
            boneDataArray!.contents().copyMemory(from: boneWeights, byteCount: weightsSize)
            boneDataArray!.contents().advanced(by: weightsSize).copyMemory(from: boneIndices, byteCount: indicesSize)
            boneDataArray!.label = mesh.displayName + "-bonedata"
            boneIndexOffset = weightsSize
        } else {
            boneDataArray = nil
            boneIndexOffset = nil
        }
    }
    
    func addToVertexArray() {
        vertexArray.add(vertices: modelMesh.vertexDataAccessors!, count: modelMesh.countOfVertices, elements: modelMesh.elementData, bytesPerElement: modelMesh.elementSize, at: reservation)
    }
    
}
